### PR TITLE
Login logout 13 pt 2

### PR DESCRIPTION
**Pull request by muydanny**
merchant new spec

Part 2 of 13 complete

Merchant login

Admin login complete

### CODE REVIEW

File: spec/features/sessions/merchant_user_new_spec.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+        expect(current_path).to eq('/login')
+
+        fill_in :email, with: @merchant_1.email
+        fill_in :password, with: @merchant_1.password
+        # how to test becrypt authentication
+
+        click_on "Login!"
+
+        expect(current_path).to eq('/')
```
Comment (muydanny) [inline spec/features/sessions/merchant_user_new_spec.rb]: THis was eventually changed dont even look at this. Its trash.

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -24,7 +24,7 @@ def show
   private
 
   def user_params
-    params.permit(:name,:address,:city,:state,:zip, :email,:password)
+    params.permit(:name, :address, :city, :state, :zip, :email, :password)
```
Comment (muydanny) [inline app/controllers/users_controller.rb:28]: OCD

File: db/seeds.rb [hunk 3]
```diff
@@ -21,6 +21,6 @@
 dog_bone = dog_shop.items.create(name: "Dog Bone", description: "They'll love it!", price: 21, image: "https://img.chewy.com/is/image/catalog/54226_MAIN._AC_SL1500_V1534449573_.jpg", active?:false, inventory: 21)
 
 #users with (role: set manually)
-default_1 = User.create(name: "Hank Hill", address: "801 N Alamo St", city: "Arlen", state: "Texas", zip: "61109", email: "ProPAIN@aol.com", password_digest: "W33dWacker", role: 0)
-merchant_1 = User.create(name: "Maude Sloggett", address: "17 Sun Rise St", city: "El Paso", state: "Illinois", zip: "56726", email: "M.Slogget@yahoo.com", password_digest: "Forever27", role: 1)
-admin_1 = User.create(name: "Kurt Cobain", address: "666 Lake Washington Bldv", city: "Seattle", state: "Washington", zip: "32786", email: "GrungeIsDead@gmail.com", password_digest: "Forever27", role: 2)
+default_1 = User.create(name: "Hank Hill", address: "801 N Alamo St", city: "Arlen", state: "Texas", zip: "61109", email: "ProPAIN@aol.com", password: "W33dWacker", role: 0)
```
Comment (muydanny) [inline db/seeds.rb:24]: This is what Michael was running into so i changed his seeds oh so slightly

File: spec/features/sessions/merchant_user_new_spec.rb [hunk 4]
```diff
@@ -3,7 +3,7 @@
 RSpec.describe "Logging in" do
     before(:each) do
         @regular_user = User.create(name: "Willy Wonka", address: "123 St", city: "Denver", state: "CO", zip: "12345", email: "chocolateguy1@gmail.com", password: "loco123", role: 0)
-        @merchant_1 = User.create(name: "Maude Sloggett", address: "17 Sun Rise St", city: "El Paso", state: "Illinois", zip: "56726", email: "M.Slogget@yahoo.com", password_digest: "Forever27", role: 1)
+        @merchant_1 = User.create(name: "Maude Sloggett", address: "17 Sun Rise St", city: "El Paso", state: "Illinois", zip: "56726", email: "M.Slogget@yahoo.com", password: "Forever27", role: 1)
```
Comment (muydanny) [inline spec/features/sessions/merchant_user_new_spec.rb]: samesies here

File: app/controllers/merchants_controller.rb [hunk 5]
```diff
@@ -5,7 +5,7 @@ def index
   end
 
   def show
-    @merchant = Merchant.find(params[:id])
+    @merchant = User.find_by(id: session[:user_id])&.merchant || Merchant.find_by(id: params[:id])
```
Comment (muydanny) [inline app/controllers/merchants_controller.rb:8]: Either we get to this route by /merchants/:id or were getting the merchant from the logged in user its associated with

File: app/controllers/sessions_controller.rb [hunk 6]
```diff
@@ -9,10 +9,41 @@ def create
...
+
+      redirect_to route
     else
       flash[:error] = "Sorry, youre not one of us."
-      render :new
     end
   end
+
+  def determine_route(user)
+    # WHEN YOU ARE CHECKING _DIFFERENT_ VALUES IN YOUR CONDITION
```
Comment (muydanny) [inline app/controllers/sessions_controller.rb:22]: The commented out code is just other ways of doing the bottom code. Ill put this elsewhere but thought maybe it was cool to compare

File: app/models/merchant.rb [hunk 7]
```diff
@@ -1,4 +1,5 @@
 class Merchant <ApplicationRecord
+  has_many :users
```
Comment (muydanny) [inline app/models/merchant.rb:2]: ok so i migrated :0 Im pretty sure in DTR we talked about getting that approved but I spaced and rolled with it. My bad. Im trash but tests are passing.

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (Gallup93) [reply app/controllers/users_controller.rb:28]: I respect it haha

Comment (stellakunzang) [reply app/controllers/users_controller.rb:28]: I LOVE IT

File: app/controllers/merchants_controller.rb (hunk 5 above)
Comment (stellakunzang) [reply app/controllers/merchants_controller.rb:8]: OOoooo look at you with the memoization!

**Pull request merged**